### PR TITLE
rfc 5861 (stale-if-error, stale-while-revalidate)

### PR DESCRIPTION
**Pull request by sithmel**
solves #27

### CODE REVIEW

**Comment by sithmel** [comment]
stale-if error works fine in cacheable-request https://github.com/lukechilds/cacheable-request/issues/94. Without changing the code. I have created a test case locally.

**Comment by kornelski** [comment]
Sorry, GitHub UI is stupid and my comments weren't public

File: index.js [hunk 1]
```diff
@@ -425,13 +446,25 @@ module.exports = class CachePolicy {
     }
 
     timeToLive() {
-        return Math.max(0, this.maxAge() - this.age()) * 1000;
+        return Math.max(
+            0,
+            this.maxAge() - this.age(),
+            this.maxAge() + toNumberOrZero(this._rescc['stale-if-error']) - this.age(),
+            this.maxAge() + toNumberOrZero(this._rescc['stale-while-revalidate']) - this.age()) * 1000;
```
[the remark this answers is not in the document]
Comment (sithmel) [reply index.js]: I have done a bit differently. I think is a bit more readable

File: index.js [hunk 2]
```diff
@@ -425,13 +446,25 @@ module.exports = class CachePolicy {
...
     }
 
     stale() {
         return this.maxAge() <= this.age();
     }
 
+    useStaleIfError() {
+        return this.maxAge() + toNumberOrZero(this._rescc['stale-if-error']) > this.age();
```
[the remark this answers is not in the document]
Comment (sithmel) [reply index.js]: done

**Comment by sithmel** [comment]
I also added an easy way to mock time. Tests were randomly breaking otherwise.

**Comment by kornelski** [comment]
Please don't add unrelated functionality to a PR.

I don't like the mocking change. The `now` method exist specifically to be monkey-patched in tests, so it doesn't need another way of being patched.

**Comment by sithmel** [comment]
I have reverted the mock. I missed to realise that there was already a way. Thanks for spotting that!

I had to monkey patch `now` for a single test that was failing in CI.

**Comment by kornelski** [comment]
Thank you

**Pull request merged**